ArduCopter/flight_mode.cpp

```diff
@@ -26,6 +26,11 @@ bool Copter::set_mode(control_mode_t mode, mode_reason_t reason)
         return true;
     }
 
+    // check if an RTL needs to be converted to an AUTO
+    if (mode == RTL && g2.rtl_type == RTLType_MissionDoLandStart) {
+        mode = check_do_land_start();
+    }
+    
```

Comment:
Actually you need _resume()_, otherwise if you have MISSION_RESTART set to 1, it will reset the mission index it just set. So just put the _mission.resume()_ inside the _check_do_land_start()_, change this to be before the previous if and you won't need the special case for _exit_mode_.